### PR TITLE
EdgeHub: Add support for a RetryingClient to handle SDK issues

### PR DESCRIPTION
**Pull request by varunpuranik**
C# SDK sometimes goes into a bad state and keeps throwing ObjectDisposedExceptions. If that happens, we need to do the following -
- Discard the existing client (DeviceClient or ModuleClient)
- Create a new one
- Retry the operation.

This PR adds such a retrying CloudProxy which will retry all the operations if the Client runs into such an error.

### CODE REVIEW

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/cloud/RetryingCloudProxy.cs [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+            for (int i = 0; i < RetryCount; i++)
+            {
+                ICloudProxy cloudProxy = await this.GetCloudProxy();
+                try
+                {
+                    return await func(cloudProxy);
+                }
+                catch (Exception)
+                {
+                    if (cloudProxy.IsActive || i + 1 == 3)
```
Comment (damonbarry) [inline edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/cloud/RetryingCloudProxy.cs]: The hard-coded `3` should probably be changed to `RetryCount`.

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/cloud/RetryingCloudProxy.cs [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+                    {
+                        throw;
+                    }
+                }
+            }
+
+            // Should never get here
+            return default(T);
```
Comment (damonbarry) [inline edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/cloud/RetryingCloudProxy.cs]: I think the compiler will allow you to remove this unreachable code if you change the `for` loop condition from `i < RetryCount` to `true`. Your `catch` clause already controls when you will exit the loop, so you don't need it twice. When you change it to `true`, the compiler knows that the _only_ way to exit the function is from inside the loop.

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/cloud/RetryingCloudProxy.cs (hunk 1 above)
Comment (damonbarry) [reply edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/cloud/RetryingCloudProxy.cs]: Already fixed!

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/cloud/RetryingCloudProxy.cs (hunk 2 above)
Comment (varunpuranik) [reply edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/cloud/RetryingCloudProxy.cs]: Good point, fixed.

**Pull request merged**